torchvision/prototype/transforms/_geometry.py

```diff
@@ -414,75 +414,92 @@ def __init__(
             _check_padding_mode_arg(padding_mode)
 
         self.padding = padding
+        self._parsed_padding = F._geometry._parse_pad_padding(padding) if padding else None  # type: ignore[arg-type]
         self.pad_if_needed = pad_if_needed
         self.fill = _setup_fill_arg(fill)
         self.padding_mode = padding_mode
 
     def _get_params(self, sample: Any) -> Dict[str, Any]:
-        _, height, width = query_chw(sample)
+        _, padded_height, padded_width = query_chw(sample)
 
-        if self.padding is not None:
-            # update height, width with static padding data
-            padding = self.padding
-            if isinstance(padding, Sequence):
-                padding = list(padding)
-            pad_left, pad_right, pad_top, pad_bottom = F._geometry._parse_pad_padding(padding)
-            height += pad_top + pad_bottom
-            width += pad_left + pad_right
+        if self._parsed_padding is not None:
+            pad_left, pad_right, pad_top, pad_bottom = self._parsed_padding
+            padded_height += pad_top + pad_bottom
+            padded_width += pad_left + pad_right
+        else:
+            pad_left = pad_right = pad_top = pad_bottom = 0
 
-        output_height, output_width = self.size
-        # We have to store maybe padded image size for pad_if_needed branch in _transform
-        input_height, input_width = height, width
+        cropped_height, cropped_width = self.size
+        height_diff = padded_height - cropped_height
+        width_diff = padded_width - cropped_width
```

Comment:
Reading all below comments and the code I still do not quite get why we could not define `height_diff = cropped_height - padded_height` such that we do not need to do `height_diff = -height_diff`

Reply:
In the sampling of `top` and `left` L485-L486 we need `padded - cropped`. So if we do `cropped - padded` here, we need to invert for the regular case if `pad_if_needed` is not set.

Since we need both definitions for the diff, we need to invert (or recompute) at least once. With the speed optimization in mind, I put the invert into the optional branch.

Reply:
Seems strange. Stable implementation does not do that. Also we have if raise clause for cases when top/left can't be sampled if difference is negative or zero...

Reply:
> Seems strange. Stable implementation does not do that.

Yes, it does. In the `pad_if_needed` branch we do

https://github.com/pytorch/vision/blob/06d9726d66f97ed8e3d3910da246a52be9c6e168/torchvision/prototype/transforms/_geometry.py#L440

and later for the `top` and `left` sampling we do

https://github.com/pytorch/vision/blob/06d9726d66f97ed8e3d3910da246a52be9c6e168/torchvision/prototype/transforms/_geometry.py#L454

> Also we have if raise clause for cases when top/left can't be sampled if difference is negative or zero...

1. If the difference is negative, we never get to the sampling and raise an error earlier the same as the stable transform.
2. If the difference is zero, we also have this branch as ternary statement in the sampling:

    ```py
    top = int(torch.randint(0, height_diff + 1, size=())) if height_diff > 0 else 0
    left = int(torch.randint(0, width_diff + 1, size=())) if width_diff > 0 else 0
    ```

Reply:
OK, i checked in details the implementation and now understand the trick with `diff=-diff`. Well, I think it would be better to keep code more transparent without this micro-optimization... I think we can afford computing once again the diff once pad_if_needed branch is applied...

Reply:
I've removed the trick in 5aff3b759, while keeping the large comment in tact for better comparability. The comment would need to be adapted if we want to keep the commit.

Removing the trick leads to three more branches and IMO does not really enhances readability. I'll let you be the judge.

Reply:
We can do something like:
```
        cropped_height, cropped_width = self.size        
                
        if self.pad_if_needed:
            if cropped_height > padded_width:
                height_diff = cropped_height - padded_height
                pad_top += height_diff
                pad_bottom += height_diff
                padded_height += 2 * height_diff

            if cropped_width > padded_width:
                width_diff = cropped_width - padded_width                        
                pad_left += width_diff
                pad_right += width_diff
                padded_width += 2 * width_diff

        height_diff = padded_height - cropped_height
        width_diff = padded_width - cropped_width
```

Reply:
Good call. Let me send a patch.